supplementary/alica_ros2_proxy/include/ros2_clock/AlicaRosTimer.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+#include "engine/AlicaClock.h"
+#include "engine/IAlicaTimer.h"
+
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+
+#include <rclcpp/duration.hpp>
+#include <rclcpp/executors.hpp>
+#include <rclcpp/node.hpp>
+
+namespace alicaRosTimer
+{
+
+class SyncStopTimerRosImpl : public std::enable_shared_from_this<SyncStopTimerRosImpl>
+{
+    using Base = std::enable_shared_from_this<SyncStopTimerRosImpl>;
+
+public:
+    using TimerCb = std::function<void()>;
+
+    SyncStopTimerRosImpl(TimerCb&& userCb, alica::AlicaTime period)
+            : _userCb(std::move(userCb))
+            , _period(toRosDuration(period))
+            , _nh(rclcpp::Node::make_shared("AlicaROS2Timer"))
+            , _timer()
+            , _stopMutex()
+            , _stopCv()
+            , _userCbInProgress(false)
+            , _stop(false)
+    {
+
+        _myexec.add_node(_nh);
+        _clock = std::thread([this]() { _myexec.spin(); });
+    }
+
+    ~SyncStopTimerRosImpl()
+    {
+        _myexec.cancel();
+        _clock.join();
+    }
+
+    void start()
+    {
+        // Call the userCb once immediately. Required since ROS timers don't fire immediately. This means the first callback
+        // will be called from the scheduler thread which is fine since stop() anyway blocks on userCb()
+        _userCb();
+
+        _timer = rclcpp::create_timer(_nh, _nh->get_clock(), _period, std::bind(&SyncStopTimerRosImpl::timerCb, this));
+    }
+
+    void stop()
+    {
+        _timer->cancel();
+        // Block if the timerCb is in progress
+        // TODO: use atomics to check if timerCb is in progress & acquire the mutex only if we need to block
+        std::unique_lock<std::mutex> lock(_stopMutex);
+        _stop = true;
+        _stopCv.wait(lock, [this]() { return !_userCbInProgress; });
+    }
+
+    void timerCb()
+    {
+        {
+            std::unique_lock<std::mutex> lock(_stopMutex);
+            if (_stop) {
+                return;
+            }
+            _userCbInProgress = true;
+        }
+
+        _userCb();
+
+        std::unique_lock<std::mutex> lock(_stopMutex);
+        _userCbInProgress = false;
+        if (_stop) {
+            _stopCv.notify_one();
```

Comment:
How would you recommend it be done instead? That code is what the original ros1 proxy had, so I didn't touch since it didn't need changing for the port to ros2